witness/golang/cmd/witness/internal/http/server_test.go

```diff
@@ -0,0 +1,380 @@
+		dh("3a357a5ff22c69641ff59c08ca67ccabdefdf317476501db8cafc73ebb5ff547", 32),
+	}
+	crNext  = []byte("Log Checkpoint v0\n15\nsrKoB8sjvP1QAt1Ih3nqGHzjvmtRLs/geQdehrUHvqs=\n\n— monkeys h74qVU+3T1AVNo+tAsISWzMwEDag8xOwQLSzxJELzHH9N/4cz6M+VbJ89Ku//gybRPPUFP/8Fcvxpqb6++ZyLk83oAE=\n")
+	crProof = [][]byte{
+		dh("ef626e0b64023948e57f34674c2574b3078c5af59a2faa095f4948736e8ca52e", 32),
+		dh("8f75f7d88d3679ac6dd5a68a81215bfbeafe8c566b93013bbc82e64295628c8b", 32),
+		dh("e034fb7af8223063c1c299ed91c11a0bc4cec15afd75e2abe4bb54c14d921ef0", 32),
+	}
+)
+
+type LogOpts struct {
+	ID         string
+	PK         string
+	useCompact bool
+}
+
+func newOpts(d *sql.DB, logs []LogOpts) (witness.Opts, error) {
+	ns, err := note.NewSigner(wSK)
+	if err != nil {
+		return witness.Opts{}, fmt.Errorf("newOpts: couldn't create a note signer: %v", err)
+	}
+	h := hasher.DefaultHasher
+	logMap := make(map[string]witness.LogInfo)
+	for _, log := range logs {
+		logV, err := note.NewVerifier(log.PK)
+		if err != nil {
+			return witness.Opts{}, fmt.Errorf("newOpts: couldn't create a log verifier")
+		}
+		sigV := []note.Verifier{logV}
+		logInfo := witness.LogInfo{
+			SigVs:      sigV,
+			Hasher:     h,
+			UseCompact: log.useCompact,
+		}
+		logMap[log.ID] = logInfo
+	}
+	opts := witness.Opts{
+		DB:        d,
+		Signer:    ns,
+		KnownLogs: logMap,
+	}
+	return opts, nil
+}
+
+func newWitness(t *testing.T, d *sql.DB, logs []LogOpts) *witness.Witness {
+	opts, err := newOpts(d, logs)
+	if err != nil {
+		t.Fatalf("couldn't create witness opt struct: %v", err)
+	}
+	w, err := witness.New(opts)
+	if err != nil {
+		t.Fatalf("couldn't create witness: %v", err)
+	}
+	return w
+}
+
+// dh is taken from https://github.com/google/trillian/blob/master/merkle/logverifier/log_verifier_test.go.
+func dh(h string, expLen int) []byte {
+	r, err := hex.DecodeString(h)
+	if err != nil {
+		panic(err)
+	}
+	if got := len(r); got != expLen {
+		panic(fmt.Sprintf("decode %q: len=%d, want %d", h, got, expLen))
+	}
+	return r
+}
+
+func mustCreateDB(t *testing.T) (*sql.DB, func() error) {
+	t.Helper()
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Fatalf("failed to open temporary in-memory DB: %v", err)
+	}
+	return db, db.Close
+}
+
+func TestGetLogs(t *testing.T) {
+	for _, test := range []struct {
+		desc       string
+		logIDs     []string
+		logPKs     []string
+		chkpts     [][]byte
+		wantStatus int
+		wantBody   string
+	}{
+		{
+			desc:       "no logs",
+			logIDs:     []string{},
+			wantStatus: http.StatusOK,
+			wantBody:   "",
+		}, {
+			desc:       "one log",
+			logIDs:     []string{"monkeys"},
+			logPKs:     []string{mPK},
+			chkpts:     [][]byte{mInit},
+			wantStatus: http.StatusOK,
+			wantBody:   `["monkeys"]`,
+		}, {
+			desc:       "two logs",
+			logIDs:     []string{"bananas", "monkeys"},
+			logPKs:     []string{bPK, mPK},
+			chkpts:     [][]byte{bInit, mInit},
+			wantStatus: http.StatusOK,
+			wantBody:   `["bananas","monkeys"]`,
```

Comment:
I wonder if it might be less brittle to have this be `wantResponse struct ...` which gets compared to a similar struct that the JSON response is parsed into?
As it is, any changes to whitespace formatting would cause the test to fail, even though it's a valid/correct response.

Reply:
Sure, changed to unmarshalling JSON and comparing the two lists.